File: core/drive/dynamics.py
"""驱力动力系统: 驱力 → 力向量 → MindState 演化。

融合旧架构的 DriveSystem + DynamicalSystem。
驱力产生力向量，心理学分析产生力向量，合力推动 MindState 更新。
"""
from __future__ import annotations

from dataclasses import dataclass

from ..mind_state import MindState
from .desires import DriveState


@dataclass
class ForceVector:
    """MindState 空间中的力向量——方向+强度。"""

    pleasure_push: float = 0.0
    arousal_push: float = 0.0
    dominance_push: float = 0.0
    control_push: float = 0.0
    attachment_push: float = 0.0
    defense_push: float = 0.0
    goal_tension_push: float = 0.0

    source: str = ""
    weight: float = 1.0

    def magnitude(self) -> float:
        return (
            abs(self.pleasure_push) + abs(self.arousal_push) +
            abs(self.dominance_push) + abs(self.control_push) +
            abs(self.attachment_push) + abs(self.defense_push) +
            abs(self.goal_tension_push)
        )

    def scale(self, factor: float) -> "ForceVector":
        return ForceVector(
            pleasure_push=self.pleasure_push * factor,
            arousal_push=self.arousal_push * factor,
            dominance_push=self.dominance_push * factor,
            control_push=self.control_push * factor,
            attachment_push=self.attachment_push * factor,
            defense_push=self.defense_push * factor,
            goal_tension_push=self.goal_tension_push * factor,
            source=self.source,
            weight=self.weight,
        )


class DriveDynamics:
    """驱力动力系统: 驱力 → 力 → MindState 演化。

    用法:
        dynamics = DriveDynamics()
        dynamics.set_baseline(ocean_profile)
        new_ms = dynamics.step(current_ms, drive_state, psychology_result, dt)
    """

    def __init__(self):
        self.damping = 0.3          # 稳定性阻尼
        self.baseline_anchor = 0.1  # OCEAN 基线锚定强度
        self.max_step_drift = 0.3   # 单步最大漂移
        self._baseline_ms: MindState | None = None

    def set_baseline(self, ocean: dict):
        """设置 OCEAN 基线——系统锚定在此附近。"""
        self._baseline_ms = MindState(ocean_baseline=dict(ocean))

    def step(
        self,
        current: MindState,
        drive_state: DriveState,
        psychology_result: dict | None = None,
        dt: float = 1.0,
    ) -> MindState:
        """一步动力演化: state(t+1) = state(t) + Σforces × (1-damping)。

        Args:
            current: 当前 MindState
            drive_state: 当前驱力状态
            psychology_result: 心理引擎输出 (PsychologyResult.mindstate)
            dt: 时间步长 (秒)

        Returns:
            新的 MindState
        """
        forces = self._compute_forces(drive_state, psychology_result)
        resultant = self._compose(forces)
        damping = self._compute_damping(current)
        new = self._apply(current, resultant, damping, dt)

        # 稳定性约束: 单步漂移不超过 max_step_drift
        dist = current.distance_to(new)
        if dist > self.max_step_drift:
            scale = self.max_step_drift / dist
            new = self._interpolate(current, new, scale)

        # 基线锚定
        if self._baseline_ms:
            new = self._anchor(new)

        new.check_stability()
        return new

    def _compute_forces(
        self, drive_state: DriveState, psychology: dict | None
    ) -> list[ForceVector]:
        """从驱力和心理分析计算力向量列表"""
        forces: list[ForceVector] = []

        # 驱力 → 力向量
        dv = drive_state.get_drive_vector()
        f_drive = ForceVector(source="drive")
        f_drive.pleasure_push = (dv.get("helpfulness", 0.5) - 0.5) * 0.3
        f_drive.attachment_push = (dv.get("connection", 0.5) - 0.5) * 0.3
        f_drive.dominance_push = (dv.get("autonomy", 0.5) - 0.5) * 0.2
        f_drive.goal_tension_push = (dv.get("achievement", 0.5) - 0.5) * 0.15
        f_drive.arousal_push = (dv.get("curiosity", 0.5) - 0.5) * 0.2
        forces.append(f_drive)

        # 心理分析 → 力向量
        if psychology:
            affect = psychology.get("affect", {})
            f_psych = ForceVector(source="psychology")
            f_psych.pleasure_push = affect.get("pleasure", 0.0) * 0.4
            f_psych.arousal_push = (affect.get("arousal", 0.5) - 0.5) * 0.3
            f_psych.dominance_push = affect.get("dominance", 0.0) * 0.3
            f_psych.attachment_push = (psychology.get("attachment_activation", 0.0) - 0.3) * 0.2
            f_psych.defense_push = (psychology.get("defense_strength", 0.0) - 0.2) * 0.2
            f_psych.control_push = (psychology.get("control", 0.5) - 0.5) * 0.3
            forces.append(f_psych)

        return forces

    def _compose(self, forces: list[ForceVector]) -> ForceVector:
        """合成所有力向量——加权平均"""
        if not forces:
            return ForceVector()
        total_weight = sum(f.weight for f in forces) or 1.0
        return ForceVector(
            pleasure_push=sum(f.pleasure_push * f.weight for f in forces) / total_weight,
            arousal_push=sum(f.arousal_push * f.weight for f in forces) / total_weight,
            dominance_push=sum(f.dominance_push * f.weight for f in forces) / total_weight,
            control_push=sum(f.control_push * f.weight for f in forces) / total_weight,
            attachment_push=sum(f.attachment_push * f.weight for f in forces) / total_weight,
            defense_push=sum(f.defense_push * f.weight for f in forces) / total_weight,
            goal_tension_push=sum(f.goal_tension_push * f.weight for f in forces) / total_weight,
            source="resultant",
        )

    def _compute_damping(self, current: MindState) -> float:
        """计算稳定性阻尼——越远离基线，阻尼越强"""
        if self._baseline_ms is None:
            return self.damping
        distance = current.distance_to(self._baseline_ms)
        return self.damping * (1.0 + distance * 2.0)

    def _apply(self, current: MindState, force: ForceVector,
               damping: float, dt: float) -> MindState:
        """将力应用到 MindState"""
        new = current.snapshot()
        factor = max(0.0, 1.0 - damping) * min(dt, 5.0) / 5.0

        new.pleasure = max(-1.0, min(1.0, current.pleasure + force.pleasure_push * factor))
        new.arousal = max(0.0, min(1.0, current.arousal + force.arousal_push * factor))
        new.dominance = max(-1.0, min(1.0, current.dominance + force.dominance_push * factor))
        new.control = max(0.0, min(1.0, current.control + force.control_push * factor))
        new.attachment_activation = max(0.0, min(1.0,
            current.attachment_activation + force.attachment_push * factor))
        new.defense_strength = max(0.0, min(1.0,
            current.defense_strength + force.defense_push * factor))
        new.goal_tension = max(0.0, min(1.0,
            current.goal_tension + force.goal_tension_push * factor))

        return new

    def _anchor(self, state: MindState) -> MindState:
        """微弱拉回基线"""
        if self._baseline_ms is None:
            return state
        anchor = self.baseline_anchor
        dist = state.distance_to(self._baseline_ms)
        if dist < 0.15:
            return state
        state.pleasure += (self._baseline_ms.pleasure - state.pleasure) * anchor
        state.control += (self._baseline_ms.control - state.control) * anchor
        return state

    @staticmethod
    def _interpolate(a: MindState, b: MindState, t: float) -> MindState:
        """在 a 和 b 之间插值"""
        result = a.snapshot()
        result.pleasure = a.pleasure + (b.pleasure - a.pleasure) * t
        result.arousal = a.arousal + (b.arousal - a.arousal) * t
        result.dominance = a.dominance + (b.dominance - a.dominance) * t
        result.control = a.control + (b.control - a.control) * t
        result.attachment_activation = a.attachment_activation + (b.attachment_activation - a.attachment_activation) * t
        result.defense_strength = a.defense_strength + (b.defense_strength - a.defense_strength) * t
        result.goal_tension = a.goal_tension + (b.goal_tension - a.goal_tension) * t
        return result

File: src/mind/dynamics.ts
import { MindState } from "../mind/state";
import { DriveState } from "./drives";

export class ForceVector {
  pleasurePush = 0; arousalPush = 0; dominancePush = 0; controlPush = 0;
  attachmentPush = 0; defensePush = 0; goalTensionPush = 0;
  source = ""; weight = 1;

  magnitude(): number {
    return Math.abs(this.pleasurePush)+Math.abs(this.arousalPush)+Math.abs(this.dominancePush)
      +Math.abs(this.controlPush)+Math.abs(this.attachmentPush)+Math.abs(this.defensePush)
      +Math.abs(this.goalTensionPush);
  }
  scale(f: number): ForceVector {
    const fv = new ForceVector();
    fv.pleasurePush=this.pleasurePush*f; fv.arousalPush=this.arousalPush*f;
    fv.dominancePush=this.dominancePush*f; fv.controlPush=this.controlPush*f;
    fv.attachmentPush=this.attachmentPush*f; fv.defensePush=this.defensePush*f;
    fv.goalTensionPush=this.goalTensionPush*f;
    fv.source=this.source; fv.weight=this.weight; return fv;
  }
}

export class DriveDynamics {
  damping = 0.3; baselineAnchor = 0.1; maxStepDrift = 0.3;
  private _baseline: MindState | null = null;

  setBaseline(ocean: Record<string,number>) {
    this._baseline = new MindState(); this._baseline.oceanBaseline = {...ocean};
  }

  step(current: MindState, ds: DriveState, psychology?: Record<string,any>|null, dt=1.0): MindState {
    const forces = this._computeForces(ds, psychology);
    const resultant = this._compose(forces);
    const damping = this._computeDamping(current);
    let n = this._apply(current, resultant, damping, dt);
    const dist = current.distanceTo(n);
    if (dist > this.maxStepDrift) n = DriveDynamics._interpolate(current, n, this.maxStepDrift/dist);
    if (this._baseline) n = this._anchor(n);
    n.checkStability();
    return n;
  }

  private _computeForces(ds: DriveState, psych?: Record<string,any>|null): ForceVector[] {
    const forces: ForceVector[] = [];
    const dv = ds.getDriveVector();
    const fd = new ForceVector(); fd.source="drive";
    fd.pleasurePush = ((dv.helpfulness??0.5)-0.5)*0.3;
    fd.attachmentPush = ((dv.connection??0.5)-0.5)*0.3;
    fd.dominancePush = ((dv.autonomy??0.5)-0.5)*0.2;
    fd.goalTensionPush = ((dv.achievement??0.5)-0.5)*0.15;
    fd.arousalPush = ((dv.curiosity??0.5)-0.5)*0.2;
    forces.push(fd);
    if (psych) {
      const aff = psych.affect??{};
      const fp = new ForceVector(); fp.source="psychology";
      fp.pleasurePush = (aff.pleasure??0)*0.4;
      fp.arousalPush = ((aff.arousal??0.5)-0.5)*0.3;
      fp.dominancePush = (aff.dominance??0)*0.3;
      fp.attachmentPush = ((psych.attachment_activation??0)-0.3)*0.2;
      fp.defensePush = ((psych.defense_strength??0)-0.2)*0.2;
      fp.controlPush = ((psych.control??0.5)-0.5)*0.3;
      forces.push(fp);
    }
    return forces;
  }

  private _compose(forces: ForceVector[]): ForceVector {
    if (!forces.length) return new ForceVector();
    const tw = forces.reduce((s,f)=>s+f.weight,0)||1;
    const r = new ForceVector(); r.source="resultant";
    r.pleasurePush=forces.reduce((s,f)=>s+f.pleasurePush*f.weight,0)/tw;
    r.arousalPush=forces.reduce((s,f)=>s+f.arousalPush*f.weight,0)/tw;
    r.dominancePush=forces.reduce((s,f)=>s+f.dominancePush*f.weight,0)/tw;
    r.controlPush=forces.reduce((s,f)=>s+f.controlPush*f.weight,0)/tw;
    r.attachmentPush=forces.reduce((s,f)=>s+f.attachmentPush*f.weight,0)/tw;
    r.defensePush=forces.reduce((s,f)=>s+f.defensePush*f.weight,0)/tw;
    r.goalTensionPush=forces.reduce((s,f)=>s+f.goalTensionPush*f.weight,0)/tw;
    return r;
  }

  private _computeDamping(current: MindState): number {
    if (!this._baseline) return this.damping;
    return this.damping * (1 + current.distanceTo(this._baseline) * 2);
  }

  private _apply(current: MindState, force: ForceVector, damping: number, dt: number): MindState {
    const n = current.clone();
    const factor = Math.max(0, 1-damping) * Math.min(dt, 5) / 5;
    n.pleasure = Math.max(-1, Math.min(1, current.pleasure + force.pleasurePush * factor));
    n.arousal = Math.max(0, Math.min(1, current.arousal + force.arousalPush * factor));
    n.dominance = Math.max(-1, Math.min(1, current.dominance + force.dominancePush * factor));
    n.control = Math.max(0, Math.min(1, current.control + force.controlPush * factor));
    n.attachmentActivation = Math.max(0, Math.min(1, current.attachmentActivation + force.attachmentPush * factor));
    n.defenseStrength = Math.max(0, Math.min(1, current.defenseStrength + force.defensePush * factor));
    n.goalTension = Math.max(0, Math.min(1, current.goalTension + force.goalTensionPush * factor));
    return n;
  }

  private _anchor(state: MindState): MindState {
    if (!this._baseline) return state;
    const dist = state.distanceTo(this._baseline);
    if (dist < 0.15) return state;
    state.pleasure += (this._baseline.pleasure - state.pleasure) * this.baselineAnchor;
    state.control += (this._baseline.control - state.control) * this.baselineAnchor;
    return state;
  }

  private static _interpolate(a: MindState, b: MindState, t: number): MindState {
    const r = a.clone();
    r.pleasure=a.pleasure+(b.pleasure-a.pleasure)*t;
    r.arousal=a.arousal+(b.arousal-a.arousal)*t;
    r.dominance=a.dominance+(b.dominance-a.dominance)*t;
    r.control=a.control+(b.control-a.control)*t;
    r.attachmentActivation=a.attachmentActivation+(b.attachmentActivation-a.attachmentActivation)*t;
    r.defenseStrength=a.defenseStrength+(b.defenseStrength-a.defenseStrength)*t;
    r.goalTension=a.goalTension+(b.goalTension-a.goalTension)*t;
    return r;
  }
}